app/controllers/resources_controller.rb

```diff
@@ -59,20 +59,48 @@ def check_permission
     privilege = params[:privilege]
     raise ArgumentError, "privilege" unless privilege
 
-    result = assumed_role.allowed_to?(privilege, resource)
+    begin
+      result = assumed_role.allowed_to?(privilege, resource)
+
+      audit_success(privilege, result)
+    rescue => e
+      audit_failure(e, privilege)
+
+      raise
+    end
 
+    head(result ? :no_content : :not_found)
+  end
+
+  def audit_success(privilege, result)
     Audit.logger.log(
       Audit::Event::Check.new(
         user: current_user,
         client_ip: request.ip,
-        resource: resource,
+        resource_id: resource.id,
         privilege: privilege,
-        role: assumed_role,
+        role_id: assumed_role.id,
         operation: "check",
         success: result
       )
     )
+  end
 
-    head(result ? :no_content : :not_found)
+  def audit_failure(err, privilege)
```

Comment:
i think we should flip the argument (i.e `(privilege, err)`) to be consistent with the `audit_success` method

Reply:
ok, working on it